File: medium/74. Search a 2D Matrix/74. Search a 2D Matrix.py
from typing import List


class Solution:
    def searchMatrix(self, matrix: List[List[int]], target: int) -> bool:

        def binary_search(arr: List[int], low: int, high: int, target: int) -> bool:
            if high >= low:
                mid = (high + low) // 2

                if arr[mid] == target:
                    return True

                elif arr[mid] > target:
                    return binary_search(arr, low, mid - 1, target)
                else:
                    return binary_search(arr, mid + 1, high, target)
            else:
                return False

        for i in range(len(matrix)):
            res = binary_search(matrix[i], 0, len(matrix[0]) - 1, target)

            if res:
                return res

        return False


solution = Solution()

print(solution.searchMatrix([[1, 3, 5, 7], [10, 11, 16, 20], [23, 30, 34, 60]], 3))
print(solution.searchMatrix([[1, 3, 5, 7], [10, 11, 16, 20], [23, 30, 34, 60]], 13))

File: medium/74. Search a 2D Matrix/74. Search a 2D Matrix.ts
function searchMatrix(matrix: number[][], target: number): boolean {
    const binarySearch = <T extends number>(arr: T[], low: T, high: T, target: T): boolean => {

        if (high >= low) {
            const middle = Math.floor((high + low) / 2)
            const num = arr[middle]

            if (num === target)
                return true

            if (num > target)
                return binarySearch(arr, low, middle - 1, target)
            else
                return binarySearch(arr, middle + 1, high, target)
        } else
            return false
    }

    for (let i = 0; i < matrix.length; i++) {
        const res = binarySearch(matrix[i], 0, matrix[i].length - 1, target)

        if (res)
            return true
    }

    return false
};